editor/plugins/tile_set_editor_plugin.cpp

```diff
@@ -1665,16 +1665,20 @@ void TileSetEditor::_on_workspace_input(const Ref<InputEvent> &p_ie) {
 
 										w.release();
 
-										undo_redo->create_action(TTR("Edit Navigation Polygon"));
-										undo_redo->add_do_method(edited_navigation_shape.ptr(), "set_vertices", polygon);
-										undo_redo->add_undo_method(edited_navigation_shape.ptr(), "set_vertices", edited_navigation_shape->get_vertices());
-										undo_redo->add_do_method(edited_navigation_shape.ptr(), "clear_polygons");
-										undo_redo->add_undo_method(edited_navigation_shape.ptr(), "clear_polygons");
-										undo_redo->add_do_method(edited_navigation_shape.ptr(), "add_polygon", indices);
-										undo_redo->add_undo_method(edited_navigation_shape.ptr(), "add_polygon", edited_navigation_shape->get_polygon(0));
-										undo_redo->add_do_method(this, "_select_edited_shape_coord");
-										undo_redo->add_undo_method(this, "_select_edited_shape_coord");
-										undo_redo->commit_action();
+										edited_navigation_shape->clear_outlines();
+										edited_navigation_shape->add_outline(polygon);
+										edited_navigation_shape->make_polygons_from_outlines();
+										// FIXME Couldn't figure out the undo_redo quagmire
```

Comment:
Maybe @KoBeWi can help?

Reply:
I don't know the exact methods, but probably something like
```C++
undo_redo->create_action(TTR("Edit Navigation Polygon"));
undo_redo->add_do_method(edited_navigation_shape.ptr(), "clear_outlines");
undo_redo->add_do_method(edited_navigation_shape.ptr(), "add_outline", polygon);
undo_redo->add_do_method(edited_navigation_shape.ptr(), "make_polygons_from_outlines");
undo_redo->add_undo_method(edited_navigation_shape.ptr(), "clear_outlines");
for (int i = 0; i < edited_navigation_shape->get_polygon_count(); i++) {
    undo_redo->add_undo_method(edited_navigation_shape.ptr(), "add_polygon", edited_navigation_shape->get_polygon(i));
}
undo_redo->commit_action();
```

Reply:
Thanks, I tested this but it seems something is still missing. It does not go back to previous state on undo only visually deletes it.

Reply:
This code looks similar to that one, maybe it can do something similar?
https://github.com/godotengine/godot/blob/1bac95b1669ed6faf17958e76a8946a88487548b/editor/plugins/tile_set_editor_plugin.cpp#L2994-L3010

Notably, it looks like the methods called on `edited_navigation_shape` don't need to be in the undo redo, it's `edited_navigation_shape` itself which is used in the `do` methods.